In [59]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import statsmodels.api as sm
import seaborn as sns
import os
import glob
import datetime

### Combining 2016 Data

In [2]:
#read the path
file_path_2016 = "../data/parking-tickets-2016"
#list all the files from the directory
file_list_2016 = os.listdir(file_path_2016)

In [3]:
df_2016 = pd.DataFrame()

# Append all files together
for file in file_list_2016:
    df_temp = pd.read_csv(f'../data/parking-tickets-2016/{file}')
    df_2016 = pd.concat([df_2016, df_temp], ignore_index=True)

### Combining 2017 Data

In [4]:
#read the path
file_path_2017 = "../data/parking-tickets-2017"
#list all the files from the directory
file_list_2017 = os.listdir(file_path_2017)

In [5]:
df_2017 = pd.DataFrame()

# Append all files together
for file in file_list_2017:
    df_temp = pd.read_csv(f'../data/parking-tickets-2017/{file}')
    df_2017 = pd.concat([df_2017, df_temp], ignore_index=True)

### Combining 2018 Data

In [6]:
#read the path
file_path_2018 = "../data/parking-tickets-2018"
#list all the files from the directory
file_list_2018 = os.listdir(file_path_2018)

In [7]:
df_2018 = pd.DataFrame()

# Append all files together
for file in file_list_2018:
    df_temp = pd.read_csv(f'../data/parking-tickets-2018/{file}')
    df_2018 = pd.concat([df_2018, df_temp], ignore_index=True)

### Combining 2019 Data

In [8]:
#read the path
file_path_2019 = "../data/parking-tickets-2019"
#list all the files from the directory
file_list_2019 = os.listdir(file_path_2019)

In [9]:
df_2019 = pd.DataFrame()

# Append all files together
for file in file_list_2019:
    df_temp = pd.read_csv(f'../data/parking-tickets-2019/{file}')
    df_2019 = pd.concat([df_2019, df_temp], ignore_index=True)

### Combining 2020 Data

In [10]:
#read the path
file_path_2020 = "../data/parking-tickets-2020"
#list all the files from the directory
file_list_2020 = os.listdir(file_path_2020)

In [11]:
df_2020 = pd.DataFrame()

# Append all files together
for file in file_list_2020:
    df_temp = pd.read_csv(f'../data/parking-tickets-2020/{file}')
    df_2020 = pd.concat([df_2020, df_temp], ignore_index=True)

### Combining 2021 Data

In [12]:
#read the path
file_path_2021 = "../data/parking-tickets-2021"
#list all the files from the directory
file_list_2021 = os.listdir(file_path_2021)

In [13]:
df_2021 = pd.DataFrame()

# Append all files together
for file in file_list_2021:
    df_temp = pd.read_csv(f'../data/parking-tickets-2021/{file}')
    df_2021 = pd.concat([df_2021, df_temp], ignore_index=True)

### Combining 2022 Data

In [14]:
#read the path
file_path_2022 = "../data/parking-tickets-2022"
#list all the files from the directory
file_list_2022 = os.listdir(file_path_2022)

In [35]:
df_2022 = pd.DataFrame()

# Append all files together
for file in file_list_2022:
    df_temp = pd.read_csv(f'../data/parking-tickets-2022/{file}')
    df_2022 = pd.concat([df_2022, df_temp], ignore_index=True)

### Merging Data Frames

In [ ]:
frames = [df_2016, df_2017, df_2018, df_2019, df_2020, df_2021, df_2022]
df_merged = pd.concat(frames)

In [37]:
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13364473 entries, 0 to 1821886
Data columns (total 11 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   tag_number_masked       object 
 1   date_of_infraction      int64  
 2   infraction_code         float64
 3   infraction_description  object 
 4   set_fine_amount         int64  
 5   time_of_infraction      float64
 6   location1               object 
 7   location2               object 
 8   location3               object 
 9   location4               object 
 10  province                object 
dtypes: float64(2), int64(2), object(7)
memory usage: 1.2+ GB


In [40]:
df_merged.shape

(13364473, 11)

#### Going to check what each column means, starting with `tag_number_masked`. Seems like they are the id number for each ticket made. Let's see if there are any unique values.

In [38]:
df_merged['tag_number_masked'].nunique()

100022

In [39]:
df_merged['tag_number_masked'].nunique()/df_merged.shape[0]*100

0.7484170905953418

#### There's 100k unique values of almost 13 million entries. Not sure how this column can help us, or what findings I may have, but it is best to leave it alone until I find out the best use for this column.

#### Next is `date_of_infraction`, we can change this to a datetime64[ns] format. But I noticed another column that gives me the time, `time_of_infraction` displayed in minutes. We can change both of these into datetime by combining them together into a new column. 

Checking for info or null values first. 

In [41]:
df_merged['date_of_infraction'].isna().sum()

0

In [102]:
df_merged['time_of_infraction'].isna().sum()

KeyError: 0

We have no null values for `date_of_infraction` but we do for `time_of_infraction`. So once we combine them we will get 9337 null values. Once we've combined them, let's get the average time for each day of the year and use that value to fill in the null values. Let's move on to changing its `dtype`.

In [88]:
df_merged['date_of_infraction'].head()

0    20161230
1    20161230
2    20161230
3    20161230
4    20161230
Name: date_of_infraction, dtype: int64

In [89]:
df_merged['date_of_infraction'] = pd.to_datetime(df_merged['date_of_infraction'], format='%Y%m%d')
df_merged['date_of_infraction'].head()

0   2016-12-30
1   2016-12-30
2   2016-12-30
3   2016-12-30
4   2016-12-30
Name: date_of_infraction, dtype: datetime64[ns]

In [47]:
df_merged['time_of_infraction'].head()

0    1637.0
1    1637.0
2    1637.0
3    1637.0
4    1637.0
Name: time_of_infraction, dtype: float64

I cannot just change this by making unit into minutes, otherwise 1637 will give me a combination of minutes instead of separating hours and minutes into 16:37:00, 
let's do a format by breaking it down to hours and minutes.

In [76]:
df_merged['hours'] = df_merged['time_of_infraction'] // 100
df_merged['minutes'] = df_merged['time_of_infraction'] % 100

In [83]:
df_merged['hours'] = pd.to_timedelta(df_merged['hours'], unit='h')
df_merged['minutes'] = pd.to_timedelta(df_merged['minutes'], unit='m')

In [84]:
# Combine date and time columns into a single datetime column
df_merged['time_of_infraction'] = (df_merged['hours'] + df_merged['minutes']).astype(str)
df_merged['time_of_infraction'] = df_merged['time_of_infraction'].astype(str).str.split().str[-1]

In [85]:
df_merged['time_of_infraction']

0          16:37:00
1          16:37:00
2          16:37:00
3          16:37:00
4          16:37:00
             ...   
1821882    09:46:00
1821883    09:47:00
1821884    09:47:00
1821885    09:47:00
1821886    09:47:00
Name: time_of_infraction, Length: 13364473, dtype: object

#### Since I have date and time, I don't think I need a necessary datetime of both, so I shall leave that separate for now, unless necessary. 

In [ ]:
#not running, non existent
df_merged['datetime_of_infraction'] = df_merged['date_of_infraction'] + df_merged['hours'] + df_merged['minutes']

Now we can see this as its own column. Best to remove `hours` and `minutes` columns.

In [92]:
df_merged.head()

tag_number_masked date_of_infraction  infraction_code  \
0          ***18342         2016-12-30            403.0   
1          ***19472         2016-12-30            403.0   
2          ***21311         2016-12-30            403.0   
3          ***25924         2016-12-30            403.0   
4          ***53833         2016-12-30            403.0   

          infraction_description  set_fine_amount time_of_infraction  \
0  STOP-SIGNED HIGHWAY-RUSH HOUR              150           16:37:00   
1  STOP-SIGNED HIGHWAY-RUSH HOUR              150           16:37:00   
2  STOP-SIGNED HIGHWAY-RUSH HOUR              150           16:37:00   
3  STOP-SIGNED HIGHWAY-RUSH HOUR              150           16:37:00   
4  STOP-SIGNED HIGHWAY-RUSH HOUR              150           16:37:00   

  location1         location2 location3 location4 province           hours  \
0        NR   1546 BLOOR ST W       NaN       NaN       ON 0 days 16:00:00   
1        NR     5418 YONGE ST       NaN       NaN       ON 0 days 16:00:00   
2       OPP    777 QUEEN ST W       NaN       NaN       ON 0 days 16:00:00   
3        NR    747 QUEEN ST E       NaN       NaN       ON 0 days 16:00:00   
4        NR  3042 DUNDAS ST W       NaN       NaN       ON 0 days 16:00:00   

          minutes  
0 0 days 00:37:00  
1 0 days 00:37:00  
2 0 days 00:37:00  
3 0 days 00:37:00  
4 0 days 00:37:00

In [94]:
df_merged.drop(columns = ['hours', 'minutes'], inplace=True)

As mentioned before, we'll have some null values in `time_of_infraction` we'd have to look into. 9337 time entries. Let's see if we can get average time by day. 

In [101]:
df_merged[df_merged['time_of_infraction'].astype(str) == '00:00:00']

tag_number_masked date_of_infraction  infraction_code  \
1430             ***51121         2016-12-31              3.0   
1431             ***22614         2016-12-31             29.0   
1432             ***40638         2016-12-31             29.0   
1433             ***70513         2016-12-31              5.0   
7023             ***65899         2016-08-27              5.0   
...                   ...                ...              ...   
1816826          ***45230         2022-12-11              9.0   
1816827          ***87751         2022-12-11              3.0   
1816828          ***96223         2022-12-11              8.0   
1816829          ***96224         2022-12-11              8.0   
1816830          ***98685         2022-12-11              5.0   

                 infraction_description  set_fine_amount time_of_infraction  \
1430           PARK ON PRIVATE PROPERTY               30           00:00:00   
1431     PARK PROHIBITED TIME NO PERMIT               30           00:00:00   
1432     PARK PROHIBITED TIME NO PERMIT               30           00:00:00   
1433     PARK-SIGNED HWY-PROHIBIT DY/TM               50           00:00:00   
7023     PARK-SIGNED HWY-PROHIBIT DY/TM               50           00:00:00   
...                                 ...              ...                ...   
1816826  STOP-SIGNED HWY-PROHIBIT TM/DY              100           00:00:00   
1816827        PARK ON PRIVATE PROPERTY               30           00:00:00   
1816828    STAND VEH.-PROHIBIT TIME/DAY              100           00:00:00   
1816829    STAND VEH.-PROHIBIT TIME/DAY              100           00:00:00   
1816830  PARK-SIGNED HWY-PROHIBIT DY/TM               50           00:00:00   

        location1            location2 location3 location4 province  
1430          NaN      35 BROOKWELL DR       NaN       NaN       ON  
1431           NR  503 PALMERSTON BLVD       NaN       NaN       ON  
1432          OPP   132 BARRINGTON AVE       NaN       NaN       ON  
1433           NR      51 BROOKWELL DR       NaN       NaN       ON  
7023           NR         36 LISGAR ST       NaN       NaN       ON  
...           ...                  ...       ...       ...      ...  
1816826        NR      42 CHARLES ST E       NaN       NaN       ON  
1816827        AT      100 COSBURN AVE       NaN       NaN       ON  
1816828        NR        55 CLINTON ST       NaN       NaN       ON  
1816829        NR        55 CLINTON ST       NaN       NaN       ON  
1816830       OPP        5 AUGUSTA AVE       NaN       NaN       ON  

[2996 rows x 11 columns]

In [31]:
# avg_time_by_day = df_2022.groupby(df_2022['datetime_of_infraction'].dt.weekday).agg({'time_of_infraction': 'mean'})
# avg_time_by_day

time_of_infraction
datetime_of_infraction                          
0.0                    0 days 11:27:47.387809911
1.0                    0 days 11:40:25.866165843
2.0                    0 days 11:28:20.732139035
3.0                    0 days 11:37:40.535615915
4.0                    0 days 11:37:30.077166685
5.0                    0 days 11:20:47.051922659
6.0                    0 days 10:53:04.571962126

We can move onto `infraction_code` and `infraction_description`.

In [32]:
# df_2022['infraction_code'].value_counts()

infraction_code
3      378837
207    284719
29     263451
5      258854
9       98682
        ...  
32          1
439         1
437         1
311         1
342         1
Name: count, Length: 160, dtype: int64

In [33]:
# df_2022['infraction_description'].value_counts()

infraction_description
PARK ON PRIVATE PROPERTY          378835
PARK MACHINE-REQD FEE NOT PAID    284719
PARK PROHIBITED TIME NO PERMIT    263450
PARK-SIGNED HWY-PROHIBIT DY/TM    258854
STOP-SIGNED HWY-PROHIBIT TM/DY     98681
                                   ...  
STOP SIGNED HWY PRO TIMES/DAYS         1
PARK-FAIL TO DEPOSIT FEE METER         1
PARK MOTORCYCLE-MCHNE LT 45DEG         1
PARK-FEE NOT PD FOR TIME USED          1
PARK LT 0.3M FROM SIDEWALK             1
Name: count, Length: 163, dtype: int64

Seems like both columns have about the same counts of each unique value, indicating the code and description aligns. Although it's not entirely accurate, let's check for null values for these columns. 

In [ ]:
# plt.figure(figsize=(80,10))
# df_2022['infraction_description'].value_counts(normalize=True).plot(kind="bar")
# plt.show()